### PR TITLE
ongoing work on painting with all predicates ablated

### PR DESCRIPTION
**Pull request by ronuchit**

### CODE REVIEW

**Comment by ronuchit** [comment]
Added a perfect debug grammar for painting, and a simple search procedure that just adds in the best-scoring next predicate repeatedly, until the score stops improving. (GBFS was getting too confusing for me...) With the debug grammar, the search decides to keep all the predicates, and then planning succeeds -- yay!

When you don't use the debug grammar, I verified that all 9 perfect predicates exist in the candidates, but the search goes astray pretty quickly. Here are the first few predicate sets in the search:
```
1. frozenset({})
2. frozenset({((0:robot).fingers<=0.5)})
3. frozenset({((0:robot).fingers<=0.5), ((0:obj).dirtiness<=0.495)})
4. frozenset({((0:robot).fingers<=0.5), ((0:obj).wetness<=0.5), ((0:obj).dirtiness<=0.495)})
5. frozenset({((0:robot).fingers<=0.5), ((0:obj).dirtiness<=0.743), ((0:obj).wetness<=0.5), ((0:obj).dirtiness<=0.495)})
6. frozenset({((0:obj).dirtiness<=0.743), ((0:obj).wetness<=0.5), ((0:obj).dirtiness<=0.495), ((0:robot).fingers<=0.5), NOT-((0:obj).pose_y<=-8.39)})
7. frozenset({((0:obj).dirtiness<=0.743), ((0:obj).wetness<=0.5), ((0:obj).dirtiness<=0.495), ((0:robot).fingers<=0.5), NOT-((0:obj).pose_y<=-8.39), NOT-((0:obj).pose_y<=-6.78)})
```

As you can see by comparing against the DebugGrammar I added, things are sailing smoothly along for 1,2,3,4 but then start to fall apart after that. Why does it select a second dirtiness predicate in 5? This needs to be investigated. I can continue tomorrow.

File: src/approaches/grammar_search_invention_approach.py [hunk 1]
```diff
@@ -34,32 +34,33 @@ def _create_grammar(dataset: Dataset, given_predicates: Set[Predicate]
...
+    # chained_grammar = _ChainPredicateGrammar([given_grammar, sfi_grammar])
     # Now, the chained grammar will undergo a series of transformations.
     # For each predicate enumerated by the chained grammar, we also
     # enumerate the negation of that predicate.
-    negated_grammar = _NegationPredicateGrammarWrapper(chained_grammar)
+    negated_grammar = _NegationPredicateGrammarWrapper(sfi_grammar)
     # For each predicate enumerated, we also enumerate foralls for
     # that predicate, along with appropriate negations.
-    forall_grammar = _ForallPredicateGrammarWrapper(negated_grammar)
+    # forall_grammar = _ForallPredicateGrammarWrapper(negated_grammar)
```
Comment (tomsilver) [inline src/approaches/grammar_search_invention_approach.py:50]: do you want to go back to the multiple grammar names?

File: src/utils.py [hunk 2]
```diff
@@ -429,6 +429,37 @@ def _run_heuristic_search(
 
     If no goal is found, returns the state with the best priority.
     """
+    cur_node: _HeuristicSearchNode[_S, _A] = _HeuristicSearchNode(
```
Comment (tomsilver) [inline src/utils.py:432]: can you make this its own search function? it seems generally useful. is it hill climbing?

**Review by tomsilver — COMMENTED** [review]
nice line of experimenting!

File: src/approaches/grammar_search_invention_approach.py (hunk 1 above)
Comment (ronuchit) [reply src/approaches/grammar_search_invention_approach.py:50]: i think it'll be better to just have flags like `include_givens_in_grammar` and `include_forall_in_grammar`

**Comment by ronuchit** [comment]
got something working, closing this PR so we can implement things properly

**Pull request closed without merging**